armi/reactor/converters/axialExpansionChanger/expansionData.py

```diff
@@ -0,0 +1,310 @@
+
+from statistics import mean
+from typing import List
+
+from armi import runLog
+from armi.materials import material
+from armi.reactor.flags import Flags
+
+TARGET_FLAGS_IN_PREFERRED_ORDER = [
+    Flags.FUEL,
+    Flags.CONTROL,
+    Flags.POISON,
+    Flags.SHIELD,
+    Flags.SLUG,
+]
+
+
+def getSolidComponents(b):
+    """
+    Return list of components in the block that have solid material.
+
+    Notes
+    -----
+    Axial expansion only needs to be applied to solid materials. We should not update
+    number densities on fluid materials to account for changes in block height.
+    """
+    return [c for c in b if not isinstance(c.material, material.Fluid)]
+
+
+class ExpansionData:
+    """Data container for axial expansion."""
+
+    def __init__(self, a, setFuel: bool, expandFromTinputToThot: bool):
+        """
+        Parameters
+        ----------
+        a: :py:class:`Assembly <armi.reactor.assemblies.Assembly>`
+            Assembly to assign component-wise expansion data to
+        setFuel: bool
+            used to determine if fuel component should be set as
+            axial expansion target component during initialization.
+            see self._isFuelLocked
+        expandFromTinputToThot: bool
+            determines if thermal expansion factors should be calculated
+            from c.inputTemperatureInC to c.temperatureInC (True) or some other
+            reference temperature and c.temperatureInC (False)
+        """
+        self._a = a
+        self.componentReferenceTemperature = {}
+        self._expansionFactors = {}
+        self._componentDeterminesBlockHeight = {}
+        self._setTargetComponents(setFuel)
+        self.expandFromTinputToThot = expandFromTinputToThot
+
+    def setExpansionFactors(self, componentLst: List, expFrac: List):
+        """Sets user defined expansion fractions.
+
+        Parameters
+        ----------
+        componentLst : List[:py:class:`Component <armi.reactor.components.component.Component>`]
+            list of Components to have their heights changed
+        expFrac : List[float]
+            list of L1/L0 height changes that are to be applied to componentLst
+
+        Raises
+        ------
+        RuntimeError
+            If componentLst and expFrac are different lengths
+        """
+        if len(componentLst) != len(expFrac):
+            runLog.error(
+                "Number of components and expansion fractions must be the same!\n"
+                f"    len(componentLst) = {len(componentLst)}\n"
+                f"        len(expFrac) = {len(expFrac)}"
+            )
+            raise RuntimeError
+        if 0.0 in expFrac:
+            msg = (
+                "An expansion fraction, L1/L0, equal to 0.0, is not physical. Expansion fractions "
+                "should be greater than 0.0."
+            )
+            runLog.error(msg)
+            raise RuntimeError(msg)
+        for exp in expFrac:
+            if exp < 0.0:
```

Comment:
Chance for a performance bump (in a refactor so out of scope, I'm not going to make this a blocker)

Can we put the check for a zeroth valued expansion fraction in the part where we're checking for negatives? List containment already iterates over the entire list. And the point about non-physical expansion factors still applies for zero and negatively valued fractions.
Something like
```python
for exp in expFrac:
    if exp <= 0:
        msg = f"Expansion fractions must be positive. An expansion fraction of {exp} is non-physical"
        raise RuntimeError(msg)
```

Reply:
@albeanth He just wants to add an equal sign. And that seems fine to me. But is there some situation in which we have zero expansion input here?  Does that ever happen as a matter of book-keeping, such that this change would break things?

Reply:
Not so much *just* adding an equal sign, but combining the
```python
if 0.0 in expFrac:
    # error with non-physical zero expansion
for exp in expFrac:
    if exp < 0:
        # error with non-physical negative expansion
```
with the snippet above. We want to error for both zero and negative expansion factors, and the error message in both cases indicates the non-physicality of a non-positive expansion factor.